lib/fs/encoder.go

```diff
@@ -0,0 +1,326 @@
+			l.Warnf("%s encoder: The %q character cannot be encoded (as it is ignored)", encoderType, string(r))
+			continue
+		}
+		if r >= encoderNumChars {
+			l.Warnf("%s encoder: The %q character cannot be encoded (as its Unicode value (%d) is above %d)", encoderType, string(r), r, encoderNumChars-1)
+			continue
+		}
+		encoded := r | encoderBaseRune
+		f.encoderRunes[r] = encoded
+
+		if !strings.ContainsRune(encodedChars, encoded) {
+			encodedChars += string(encoded)
+		}
+	}
+}
+
+func (f *baseEncoder) CharsToEncode() string {
+	return f.charsToEncode
+}
+
+func (f *baseEncoder) EncoderRunes() [encoderNumChars]rune {
+	return f.encoderRunes
+}
+
+func (f *baseEncoder) EncoderType() FilesystemEncoderType {
+	return f.encoderType
+}
+
+func (f *baseEncoder) AllowReservedFilenames() bool {
+	return f.allowReservedFilenames
+}
+
+// path is a complete path with pathSeparator characters.
+func (f *baseEncoder) encode(path string) string {
+	encoded := ""
+
+	// We may want to use
+	// strings.Split(filepath.Clean(path), pathSeparatorString)
+	// instead.
+	for i, part := range PathComponents(path) {
+		if i > 0 {
+			encoded += pathSeparatorString
+		}
+		if part != "" {
+			encoded += f.encodeName(part)
+		}
+	}
+
+	return encoded
+}
+
+// name is a file or directory name without pathSeparator characters.
+func (f *baseEncoder) encodeName(name string) string {
+	if name == "" || name == "." || name == ".." {
+		return name
+	}
+
+	return string(f.encodeToRunes(name))
+}
+
+// name is a file or directory name without pathSeparator characters.
+func (f *baseEncoder) encodeToRunes(name string) []rune {
+	if name == "" || name == "." || name == ".." {
+		return []rune(name)
+	}
+
+	runes := []rune(name)
+	for i, r := range runes {
+		if r >= 0 && r < encoderNumChars {
+			runes[i] = f.encoderRunes[r]
+		}
+	}
+
+	return runes
+}
+
+func (f *baseEncoder) filter(files []string) []string {
+	// Only the standard encoder filters encoded files.
+	if f.encoderType != FilesystemEncoderTypeStandard {
+		return files
+	}
+	var filtered []string
+	for _, path := range files {
+		if isEncoded(path) {
+			if l.IsTraced("encoder") {
+				l.Debugf(filenameContainsSyncthingReservedChars, path)
+			}
+			continue
+		}
+		filtered = append(filtered, path)
+	}
+	return filtered
+}
+
+// path is a complete path with pathSeparator characters.
+func (f *baseEncoder) wouldEncode(path string) bool {
+	return f.encode(path) != path
+}
+
+// decode returns the decoded path. It doesn't matter what encoder is
+// being used, as they all decode using the same algorithm.
+// We only decode those characters that any encoder encodes.
+//
+// We could choose to decode all characters in the \uf000-\uf0ff space,
+// as all of these characters are reserved for Syncthing's use, and
+// future encoders may use any of these characters, but we don't
+// currently. It seems it's best to only decode what we encode per
+// https://github.com/microsoft/WSL/issues/3200#issuecomment-389613611
```

Comment:
You say "reserved for Syncthing" but that's not really the case is it, given that there are a bunch of products apparently out there using this encoding as well? Isn't there a risk of significant weirdness if you for example sync a folder in Windows where WSL has written encoded file names?

Reply:
Yes, it would be best if the user just ran Syncthing inside their WSL instance. That's effectively what this PR trying to do, workaround the limitations of the filesystem.